client.go

```diff
@@ -1531,6 +1531,101 @@ func (cl *Client) AddTorrentInfoHash(infoHash metainfo.Hash) (t *Torrent, new bo
 	return
 }
 
+// Save torrent to state file, metadata + pices states.
+func (cl *Client) SaveTorrent(t *Torrent) []byte {
+	t.cl.mu.Lock()
+	defer t.cl.mu.Unlock()
+
+	return nil
+}
+
+// Load torrent from saved state, metadata + pices states.
+func (cl *Client) LoadTorrent(buf []byte) (t *Torrent, err error) {
+	r := bytes.NewReader(buf)
+	mi, err := metainfo.Load(r)
+	if err != nil {
+		return
+	}
+	var pieces []bool
+	d := bencode.NewDecoder(r)
+	err = d.Decode(&pieces)
+	if err != nil {
+		return
+	}
+	spec := TorrentSpecFromMetaInfo(mi)
+	t, _ = cl.AddTorrentInfoHash(spec.InfoHash)
+	if spec.DisplayName != "" {
+		t.SetDisplayName(spec.DisplayName)
+	}
+	if spec.Info != nil {
+		err = t.loadTorrent(spec.Info.Bytes, pieces)
+		if err != nil {
+			return
+		}
+	}
+	cl.mu.Lock()
+	defer cl.mu.Unlock()
+	if spec.ChunkSize != 0 {
+		t.chunkSize = pp.Integer(spec.ChunkSize)
+	}
+	t.addTrackers(spec.Trackers)
+	return
+}
+
+// Separate StartTorrent action from adding torrent, so we can add torrent, run
+// file consictency checks and then manually start it.
+func (cl *Client) StartTorrent(t *Torrent) {
+	if t.closed.IsSet() {
+		// reset close
+		t.closed.Clear()
+
+		t.cl.mu.Lock()
+		defer t.cl.mu.Unlock()
+
+		// reopen storage
+		info := t.info
+		t.info = nil
+		t.setInfoBytes(info.Bytes)
+	}
+
+	t.maybeNewConns()
+}
+
+// Drop connections from a torrent, prepare to Remove torrent from client or
+// Resume it again. Stop command also cancels check() torrent.
+func (cl *Client) StopTorrent(t *Torrent) {
+	t.close()
+}
+
+// Run file consistency checks. For active torrent, pause it, then resume
+// download. For Paused torrent keep it paused.
+func (cl *Client) CheckTorrent(t *Torrent) {
+	restart := false
+	if !t.closed.IsSet() {
+		t.close()
+		restart = true
+	}
+
+	if t.closed.IsSet() {
+		// reset close
+		t.closed.Clear()
+
+		t.cl.mu.Lock()
+		defer t.cl.mu.Unlock()
+
+		// reopen storage
+		info := t.info
+		t.info = nil
+		t.setInfoBytes(info.Bytes)
+	}
+
+	t.check(func() {
+		if restart {
+			cl.StartTorrent(t)
+		}
+	})
+}
```

Comment:
This may have some value. If user code knows that the data state may be incorrect or have changed considerably, we may want to force a recheck without opening a reader and scanning through the whole torrent ourselves.